extensions/amp-fit-text/0.2/fit-text.js

```diff
@@ -0,0 +1,163 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import * as styles from './fit-text.css';
+import {omit} from '../../../src/utils/object';
+import {px, setStyle} from '../../../src/style';
+import {useCallback, useEffect, useRef, useState} from '../../../src/preact';
+
+const {LINE_HEIGHT_EM_} = styles;
+
+/**
+ * @param {!JsonObject} props
+ * @return {PreactDef.Renderable}
+ */
+export function FitText(props) {
+  const {
+    'children': children,
+    'minFontSize': minFontSize = 6,
+    'maxFontSize': maxFontSize = 72,
+    'width': width = props['style']['width'] || '100%',
+    'height': height = props['style']['height'] || '100%',
+  } = props;
+  const rest = omit(props, [
+    'children',
+    'minFontSize',
+    'maxFontSize',
+    'width',
+    'height',
+  ]);
+  const contentRef = useRef(null);
+  const measurerRef = useRef(null);
+  const [wrapperStyle, setWrapperStyle] = useState(/** @type {Object} */ ({}));
+
+  const resizer = useCallback(() => {
+    return (maxHeight, maxWidth) => {
+      if (!measurerRef.current) {
+        return {};
+      }
+      const fontSize = calculateFontSize(
+        measurerRef.current,
+        Number(maxHeight),
+        Number(maxWidth),
+        minFontSize,
+        maxFontSize
+      );
+      const overflownStyle = getOverflowStyle(
+        measurerRef.current,
+        maxHeight,
+        fontSize
+      );
+      return {...overflownStyle, fontSize: px(fontSize)};
+    };
+  }, [maxFontSize, minFontSize]);
+
+  // Font size should readjust when container resizes.
+  useEffect(() => {
+    const node = contentRef.current;
+    const observer = new ResizeObserver((entries) => {
+      const last = entries[entries.length - 1];
+      const {height: maxHeight, width: maxWidth} = last['contentRect'];
+      setWrapperStyle(resizer()(maxHeight, maxWidth));
+    });
+    if (node) {
```

Comment:
Nit: Can we check the `node` earlier and skip creating the ResOb if there's not one? Might speed up the initial render.